### PR TITLE
ttl: add `ttl_job_interval` attribute and remove `tidb_ttl_job_interval` variable

### LINKED ISSUE #40028: ttl job scheduler configuration per table
Now we are using a single global variable `tidb_ttl_job_run_interval` to control the interval between TTL jobs. We could add an attribute `ttl_cronjob_schedule` to control the interval per table.

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: YangKeao <yangkeao@chunibyo.icu>

### What problem does this PR solve?

Issue Number: close #40028

Problem Summary:

The global config `tidb_ttl_job_interval` is not flexible enough. We could add an attribute for every table to configure different scan / delete frequency.

### What is changed and how it works?

Add a new attribute `ttl_job_interval` to the table, and use this attribute to decide whether this job could be scheduled.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
Allow configuring different ttl job frequency for tables.
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
@YangKeao  Please fix the linter.

```
parser/model/model.go:1755:6: exported: exported type TTLCronJobSchedule should have comment or be unexported (revive)

parser/model/model.go:1758:2: exported: exported const TTLCronJobScheduleHourly should have comment (or a comment on this block) or be unexported (revive)

parser/model/model.go:1762:1: exported: exported method TTLCronJobSchedule.Duration should have comment or be unexported (revive)

```

**Comment by YangKeao** [comment]
> @YangKeao Please fix the linter.
> 
> ```
> parser/model/model.go:1755:6: exported: exported type TTLCronJobSchedule should have comment or be unexported (revive)
> 
> parser/model/model.go:1758:2: exported: exported const TTLCronJobScheduleHourly should have comment (or a comment on this block) or be unexported (revive)
> 
> parser/model/model.go:1762:1: exported: exported method TTLCronJobSchedule.Duration should have comment or be unexported (revive)
> ```

Fixed. PTAL

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -5382,11 +5390,17 @@ func (d *ddl) AlterTableTTLInfoOrEnable(ctx sessionctx.Context, ident ast.Ident,
...
+			if ttlEnable != nil && ttlCronJobSchedule == nil {
+				return errors.Trace(dbterror.ErrSetTTLOptionForNonTTLTable.FastGenByArgs("TTL_ENABLE"))
+			} else if ttlEnable == nil && ttlCronJobSchedule != nil {
+				return errors.Trace(dbterror.ErrSetTTLOptionForNonTTLTable.FastGenByArgs("TTL_CRONJOB_SCHEDULE"))
+			} else {
+				return errors.Trace(dbterror.ErrSetTTLOptionForNonTTLTable.FastGenByArgs("TTL_ENABLE and TTL_CRONJOB_SCHEDULE"))
+			}
 		}
 
 		job = &model.Job{
```
Comment (lcwangchao) [inline ddl/ddl_api.go:5406]: Seems we can move the `job = &model.Job{...` outside the block to reduce some dup codes.

**Review by lcwangchao — APPROVED** [review]

File: ddl/ddl_api.go (hunk 1 above)
Comment (YangKeao) [reply ddl/ddl_api.go]: Done.

File: ttl/ttlworker/job_manager.go [hunk 2]
```diff
@@ -477,30 +477,34 @@ tblLoop:
...
-	if table.LastJobFinishTime.IsZero() {
+	if tableStatus.LastJobFinishTime.IsZero() {
 		return true
 	}
 
-	finishTime := table.LastJobFinishTime
+	finishTime := tableStatus.LastJobFinishTime
 
-	return finishTime.Add(variable.TTLJobRunInterval.Load()).Before(now)
+	return finishTime.Add(table.TTLInfo.CronJobSchedule.Duration()).Before(now)
```
Comment (bb7133) [inline ttl/ttlworker/job_manager.go:507]: Could you confirm if this is the expected behavior? If we follow the [CRON syntax](https://cron.help/), that would mean that it starts 'at the start of each hour/day'.

Comment (YangKeao) [reply ttl/ttlworker/job_manager.go:507]: It's not a CRON syntax. Actually it only supports "@hourly" and "@daily", one returns `time.Hour`, the other returns `24 * time.Hour`.

I've discussed this behavior with @SunRunAway. The feedback is that we only need "@hourly" and "@daily", so don't provide more complex options for users.

**Review by hawkingrei — APPROVED** [review]

File: parser/duration/duration.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	"strconv"
+	"time"
+	"unicode"
+
+	"github.com/pingcap/errors"
+)
+
+// Duration is a format which supports 'd', 'h' and 'm', but doesn't support other formats and floats
+type Duration struct {
+	Day    int `json:"day"`
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply parser/duration/duration.go:27]: Because it doesn't support 'd'

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -3031,7 +3032,7 @@ func SetDirectPlacementOpt(placementSettings *model.PlacementSettings, placement
 
 // handleTableOptions updates tableInfo according to table options.
 func handleTableOptions(options []*ast.TableOption, tbInfo *model.TableInfo) error {
-	var handledTTLOrTTLEnable bool
+	var handledTTLOrTTLOptions bool
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply ddl/ddl_api.go]: Done

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -5424,29 +5433,26 @@ func (d *ddl) AlterTableTTLInfoOrEnable(ctx sessionctx.Context, ident ast.Ident,
...
-			SchemaName: schema.Name.L,
-			TableName:  tableName,
-			Type:       model.ActionAlterTTLInfo,
-			BinlogInfo: &model.HistoryInfo{},
-			Args:       []interface{}{ttlInfo, ttlEnable},
-		}
 	} else {
 		if tblInfo.TTLInfo == nil {
-			return errors.Trace(dbterror.ErrSetTTLEnableForNonTTLTable)
+			if ttlEnable != nil && ttlCronJobSchedule == nil {
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply ddl/ddl_api.go]: Done

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**